File: python-legacy/app/services/search_service.py
"""Service for ABAP object search operations."""

import logging
from xml.etree import ElementTree as et
from typing import List, Dict

from app.core.rfc_adapter import RfcAdapter
from app.services.base_service import BaseService

logger = logging.getLogger(__name__)

# XML Namespaces for parsing ADT responses
XML_NAMESPACES = {
    "chkl": "http://www.sap.com/abapxml/checklist",
    "atom": "http://www.w3.org/2005/Atom",
    "adtcore": "http://www.sap.com/adt/core",
    "exc": "http://www.sap.com/abapxml/types/communicationframework",
    "asx": "http://www.sap.com/abapxml",
    "aunit": "http://www.sap.com/adt/aunit",
    "chkrun": "http://www.sap.com/adt/checkrun",
    "abapsource": "http://www.sap.com/adt/abapsource",
}


class SearchService(BaseService):
    """Service for searching ABAP objects via RFC."""

    def search_objects(self, query: str, max_results: int = 10) -> List[Dict[str, str]]:
        """
        Search for ABAP objects using quick search.

        Args:
            query: Search query (object name pattern, e.g., 'Z*', 'ZTEST*')
            max_results: Maximum number of results to return

        Returns:
            List[Dict[str, str]]: List of matching objects with their attributes

        Raises:
            Exception: If the request fails

        Example:
            >>> service = SearchService(connection)
            >>> results = service.search_objects("ZTEST*", max_results=5)
            >>> for obj in results:
            ...     print(f"{obj['name']} - {obj['type']}")
        """
        uri = "/sap/bc/adt/repository/informationsystem/search"

        logger.info(f"Searching for objects matching: {query}")

        with self._get_adapter() as adapter:
            response = adapter.request(
                uri=uri,
                method="GET",
                params={"operation": "quickSearch", "query": query, "maxResults": max_results},
                body="",
            )

        if response.status_code == 200:
            elements = self._parse_search_results(response.text)
            logger.debug(f"Found {len(elements)} objects matching {query}")
            return elements
        else:
            error_msg = f"{response.status_code} - Failed to search objects"
            logger.error(error_msg)
            raise Exception(f"{error_msg}\n{response.text}")

    def _parse_search_results(self, xml_text: str) -> List[Dict[str, str]]:
        """
        Parse search results from XML response.

        Args:
            xml_text: XML response text

        Returns:
            List[Dict[str, str]]: List of object attributes
        """
        root = et.fromstring(xml_text)
        elements = root.findall("adtcore:objectReference", XML_NAMESPACES)

        processed_elements = []
        for element in elements:
            cleaned = self._element_to_dict(element)
            processed_elements.append(cleaned)

        return processed_elements

    def _element_to_dict(self, element: et.Element) -> Dict[str, str]:
        """
        Convert XML element attributes to dictionary with cleaned keys.

        Args:
            element: XML element

        Returns:
            Dict[str, str]: Dictionary of attributes with namespace stripped
        """
        cleaned = {}
        for key, value in element.attrib.items():
            # Strip namespace from attribute name
            clean_key = self._strip_namespace(key)
            cleaned[clean_key] = value
        return cleaned

    def _strip_namespace(self, name: str) -> str:
        """
        Strip namespace from XML tag or attribute name.

        Args:
            name: Tag or attribute name with namespace

        Returns:
            str: Name without namespace
        """
        if name.startswith("{"):
            return name.split("}", 1)[1]
        if ":" in name:
            return name.split(":", 1)[1]
        return name

File: src/main/java/com/crystal/mcp/sapserver/service/SearchService.java
package com.crystal.mcp.sapserver.service;

import com.crystal.mcp.sapserver.model.SearchResult;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.NodeList;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import java.io.ByteArrayInputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Service for ABAP repository search operations.
 *
 * This service implements Progressive Discovery Stage 1: Quick Search.
 * Returns lightweight object references without fetching full source code.
 *
 * Thread Safety: Stateless service, thread-safe via RfcAdapter.
 *
 * Supported operations:
 * - Quick search for ABAP objects by keyword
 *
 * Future operations (Progressive Discovery):
 * - Advanced search with filters
 * - Fuzzy search
 * - Search by object type
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class SearchService {

    private final RfcAdapter rfcAdapter;

    /**
     * Search for ABAP objects using quick search.
     *
     * This method implements Progressive Discovery Stage 1:
     * - Returns lightweight results (names, types, URIs)
     * - No source code fetched (saves tokens)
     * - User can then use get_object_structure or get_object_source for details
     *
     * ADT API Endpoint:
     * GET /sap/bc/adt/repository/informationsystem/search
     * ?operation=quickSearch&query={query}&maxResults={maxResults}
     *
     * XML Response Example:
     * <objectReferences xmlns="http://www.sap.com/adt/quicksearch">
     *   <objectReference adtcore:uri="/sap/bc/adt/oo/classes/zcl_test"
     *                    adtcore:type="CLAS/OC"
     *                    adtcore:name="ZCL_TEST"
     *                    adtcore:packageName="ZTEST">
     *     <adtcore:description>Test Class</adtcore:description>
     *   </objectReference>
     * </objectReferences>
     *
     * @param query      search keyword (e.g., "ZCL_", "payment", "*invoice*")
     * @param maxResults maximum results to return (default: 10, max: 100)
     * @return SearchResult containing list of matching objects
     * @throws RuntimeException if search fails
     */
    public SearchResult searchObjects(String query, Integer maxResults) {
        // Set default max results
        int actualMaxResults = (maxResults != null && maxResults > 0)
                ? Math.min(maxResults, 100)
                : 10;

        // Build ADT API URI
        String uri = "/sap/bc/adt/repository/informationsystem/search";

        // Query parameters
        Map<String, String> params = new HashMap<>();
        params.put("operation", "quickSearch");
        params.put("query", query);
        params.put("maxResults", String.valueOf(actualMaxResults));

        log.info("Searching objects: query='{}', maxResults={}",
                query, actualMaxResults);

        try {
            // Execute RFC request
            RfcAdapter.RfcResponse response = rfcAdapter.request(
                    uri,
                    "GET",
                    null,
                    params,
                    "",
                    "application/xml"
            );

            // Check HTTP status
            if (response.statusCode() == 200) {
                // Log raw XML response for debugging
                log.debug("Raw XML response (first 500 chars): {}",
                        response.text().substring(0, Math.min(500, response.text().length())));

                // Parse XML response
                List<SearchResult.ObjectReference> results = parseSearchResults(response.text());

                log.info("Found {} objects for query '{}'", results.size(), query);

                return new SearchResult(
                        query,
                        actualMaxResults,
                        results.size(),  // Note: ADT API doesn't return total count
                        results
                );
            } else {
                String errorMsg = String.format(
                        "Search failed: HTTP %d - %s",
                        response.statusCode(),
                        response.text()
                );
                log.error(errorMsg);
                throw new RuntimeException(errorMsg);
            }

        } catch (Exception e) {
            log.error("Error searching objects for query '{}': {}",
                    query, e.getMessage(), e);
            throw new RuntimeException("Failed to search objects", e);
        }
    }

    /**
     * Parse XML search results from ADT API response.
     *
     * Handles namespaces:
     * - xmlns="http://www.sap.com/adt/quicksearch"
     * - xmlns:adtcore="http://www.sap.com/adt/core"
     *
     * @param xmlText XML response body
     * @return list of ObjectReference objects
     */
    private List<SearchResult.ObjectReference> parseSearchResults(String xmlText) {
        List<SearchResult.ObjectReference> results = new ArrayList<>();

        try {
            // Parse XML
            DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
            factory.setNamespaceAware(true);  // Important for namespace handling

            DocumentBuilder builder = factory.newDocumentBuilder();
            Document doc = builder.parse(
                    new ByteArrayInputStream(xmlText.getBytes(StandardCharsets.UTF_8))
            );

            // Define namespaces
            String ADTCORE_NS = "http://www.sap.com/adt/core";

            // Find all objectReference elements
            NodeList objectRefs = doc.getElementsByTagNameNS("*", "objectReference");

            for (int i = 0; i < objectRefs.getLength(); i++) {
                Element objRef = (Element) objectRefs.item(i);

                // Extract attributes from adtcore namespace
                String name = objRef.getAttributeNS(ADTCORE_NS, "name");
                String type = objRef.getAttributeNS(ADTCORE_NS, "type");
                String uri = objRef.getAttributeNS(ADTCORE_NS, "uri");
                String packageName = objRef.getAttributeNS(ADTCORE_NS, "packageName");

                // Extract description from child element
                String description = "";
                NodeList descNodes = objRef.getElementsByTagNameNS(ADTCORE_NS, "description");
                if (descNodes.getLength() > 0) {
                    description = descNodes.item(0).getTextContent();
                }

                // Create ObjectReference
                SearchResult.ObjectReference ref = new SearchResult.ObjectReference(
                        name != null ? name : "",
                        type != null ? type : "",
                        uri != null ? uri : "",
                        description != null ? description.trim() : "",
                        packageName != null ? packageName : ""
                );

                results.add(ref);

                log.debug("Found object: {} ({}) in package {}",
                        name, type, packageName);
            }

        } catch (Exception e) {
            log.error("Failed to parse search results XML: {}", e.getMessage(), e);
            throw new RuntimeException("XML parsing error", e);
        }

        return results;
    }
}
